tensorflow/compiler/xla/service/gpu/cudnn_fused_mha_rewriter.cc

```diff
@@ -0,0 +1,368 @@
+    if (VLOG_IS_ON(2)) {
+      VLOG(2) << "BMM1 lhs_non_contracting_dims: "
+              << absl::StrJoin(lhs_non_contracting_dims_bmm1, ",")
+              << " BMM1 rhs_non_contracting_dims: "
+              << absl::StrJoin(rhs_non_contracting_dims_bmm1, ",")
+              << " are not supported. The non-contracting dims should be less "
+                 "than 512. This is a criteria for current cuDNN 8.8 support.";
+    }
+    return false;
+  }
+
+  std::vector<int64_t> lhs_contracting_dims_bmm1 =
+      GetDimensionVector(bmm_1->operand(0)->shape().dimensions(),
+                         dot_dims_bmm1.lhs_contracting_dimensions());
+  std::vector<int64_t> rhs_contracting_dims_bmm1 =
+      GetDimensionVector(bmm_1->operand(1)->shape().dimensions(),
+                         dot_dims_bmm1.rhs_contracting_dimensions());
+
+  // The contracting dimensions for BMM1 (both lhs and rhs) need to be 64 and should be the fastest moving dimension.
+  absl::Span<const int64_t> rhs_minor_to_major_bmm1 =
+      bmm_1->operand(1)->shape().layout().minor_to_major();
+  absl::Span<const int64_t> lhs_minor_to_major_bmm1 =
+      bmm_1->operand(0)->shape().layout().minor_to_major();
+  if (!IsContractingDimSupported(lhs_contracting_dims_bmm1) ||
+      !IsDimensionMostMinor(lhs_minor_to_major_bmm1,
+                            dot_dims_bmm1.lhs_contracting_dimensions()[0]) ||
+      !IsContractingDimSupported(rhs_contracting_dims_bmm1) ||
+      !IsDimensionMostMinor(rhs_minor_to_major_bmm1,
+                            dot_dims_bmm1.rhs_contracting_dimensions()[0])) {
+    if (VLOG_IS_ON(2)) {
+      VLOG(2) << "BMM1 lhs_contracting_dims: "
+              << bmm_1->operand(0)->shape().ToString(/*print layout*/ true)
+              << " BMM1 rhs_contracting_dims: "
+              << bmm_1->operand(1)->shape().ToString(true)
+              << " are not supported. Either contracting dim is not equal to "
+                 "64 or they are not the fastest moving. This is a criteria "
+                 "for current cuDNN 8.8 supports";
+    }
+    return false;
+  }
+  return true;
+}
+
+StatusOr<bool> IsSupportedBMM2(const HloInstruction* bmm_2) {
+  GemmBackendConfig config_bmm2;
+  TF_ASSIGN_OR_RETURN(config_bmm2, bmm_2->backend_config<GemmBackendConfig>());
+  const DotDimensionNumbers& dot_dims_bmm2 =
+      config_bmm2.dot_dimension_numbers();
+
+  TF_ASSIGN_OR_RETURN(
+      std::vector<int64_t> rhs_non_contracting_dim_nums_bmm2,
+      GetNonContractingDims(bmm_2->operand(1)->shape(),
+                            dot_dims_bmm2.rhs_batch_dimensions(),
+                            dot_dims_bmm2.rhs_contracting_dimensions()));
+
+  std::vector<int64_t> rhs_non_contracting_dims_bmm2 =
+      GetDimensionVector(bmm_2->operand(1)->shape().dimensions(),
+                         rhs_non_contracting_dim_nums_bmm2);
+  // The non contracting dimension for BMM2 needs to be 64 for the input matrix.
```

Comment:
What's the significance of 64? Is that how fMHA looks in practice in LLMs?

Reply:
64 relates to the L1 cache and/or shared memory constraint in the kernel implemented by the cuDNN team. I'll find a more comprehensive answer for this. FMHA in current LLMs (certainly T5x) do look like this. So we address the immediate goal of improving perf of current crop of LLMs.